rand_core/src/lib.rs

```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
     __Nonexhaustive,
 }
 
-#[cfg(feature="std")]
+impl ErrorKind {
+    pub fn description(&self) -> &'static str {
+        match *self {
+            ErrorKind::Unavailable => "permanent failure or unavailable",
+            ErrorKind::Transient => "transient failure",
+            ErrorKind::NotReady => "not ready yet",
+            ErrorKind::Other => "uncategorised error",
+            ErrorKind::__Nonexhaustive => unreachable!(),
+        }
+    }
+}
+
 #[derive(Debug)]
-pub struct Error {
-    pub kind: ErrorKind,
-    pub cause: Option<Box<std::error::Error>>,
+enum ErrorDetails {
+    None,
+    Str(&'static str),
+    #[cfg(feature="std")]
+    Error(Box<std::error::Error + Send + Sync>),
 }
 
-#[cfg(not(feature="std"))]
-#[derive(Debug, Clone, PartialEq, Eq)]
+/// Error type of random number generators
+/// 
+/// This embeds two things, an `ErrorKind`, which can be matched over, and
+/// optional details provided to `new_str` or `new_err`.
+#[derive(Debug)]
 pub struct Error {
+    /// Error kind
     pub kind: ErrorKind,
-    pub cause: Option<&'static str>,
+    details: ErrorDetails,
 }
 
 impl Error {
+    /// Create a new instance, with no details.
+    pub fn new(kind: ErrorKind) -> Self {
+        Self { kind, details: ErrorDetails::None }
+    }
+    
+    /// Create a new instance, with a static string describing the details.
+    /// 
+    /// This may be used in both `std` and `no_std` enviroments.
+    pub fn new_str(kind: ErrorKind, details: &'static str) -> Self {
```

Comment:
The name could be a little bit confusing. Maybe a more verbose `new_with_cause_str`?

Reply:
Why the long name?

Reply:
I agree with `new_str` being a bit confusing, but I personally can't come up with the better alternative, so I guess we can leave it up to docs.